core/src/main/java/com/scalar/db/transaction/consensuscommit/TwoPhaseConsensusCommit.java

```diff
@@ -0,0 +1,362 @@
+    }
+  }
+
+  /**
+   * Retrieves results from the storage through a transaction with the specified {@link Scan}
+   * command with a partition key and returns a list of {@link Result}. Results can be filtered by
+   * specifying a range of clustering keys. Note that the current implementation clears the
+   * specified projections in {@link Scan} to project all the values including metadata properly,
+   * but the behavior might be changed at some point.
+   *
+   * @param scan a {@code Scan} command
+   * @return a list of {@link Result}
+   * @throws CrudException if the operation failed
+   */
+  @Override
+  public List<Result> scan(Scan scan) throws CrudException {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    setTargetToIfNot(scan);
+    scan.clearProjections(); // project all
+    try {
+      return crud.scan(scan);
+    } catch (UncommittedRecordException e) {
+      lazyRecovery(scan, e.getResults());
+      throw e;
+    }
+  }
+
+  @Override
+  public void put(Put put) {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    putInternal(put);
+  }
+
+  @Override
+  public void put(List<Put> puts) {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    checkArgument(puts.size() != 0);
+    puts.forEach(this::putInternal);
+  }
+
+  private void putInternal(Put put) {
+    setTargetToIfNot(put);
+    crud.put(put);
+  }
+
+  @Override
+  public void delete(Delete delete) {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    deleteInternal(delete);
+  }
+
+  @Override
+  public void delete(List<Delete> deletes) {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    checkArgument(deletes.size() != 0);
+    deletes.forEach(this::deleteInternal);
+  }
+
+  private void deleteInternal(Delete delete) {
+    setTargetToIfNot(delete);
+    crud.delete(delete);
+  }
+
+  @Override
+  public void mutate(List<? extends Mutation> mutations) {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    updateTransactionExpirationTime();
+    checkArgument(mutations.size() != 0);
+    mutations.forEach(
+        m -> {
+          if (m instanceof Put) {
+            putInternal((Put) m);
+          } else if (m instanceof Delete) {
+            deleteInternal((Delete) m);
+          }
+        });
+  }
+
+  @Override
+  public void prepare() throws PrepareException {
+    checkStatus("The transaction is not active", Status.ACTIVE);
+    beforePrepareHook.run();
+    updateTransactionExpirationTime();
+
+    try {
+      commit.prepare(crud.getSnapshot(), false);
+      status = Status.PREPARED;
+    } catch (CommitConflictException e) {
+      status = Status.PREPARE_FAILED;
+      throw new PrepareConflictException("prepare failed", e);
```

Comment:
Yes, I thought of the same thing but in the current exception structure, I think we need to do that to keep the backward compatibility.

I think introducing ConflictException will break the compatibility because that will change the exception signature of the CommitHandler methods. So I don't think we should do that at this time and maybe we can to that in the next major version.

Reply:
As discussed, we won't introduce ConflictException at this time. Maybe we will do that in the next major version.